nestkernel/eprop_archiving_node.cpp

```diff
@@ -0,0 +1,300 @@
+    *finish = it_first + std::max( 0, pos_t2 );
+  }
+}
+
+void
+nest::EpropArchivingNode::get_eprop_history( double t1, std::deque< histentry_eprop >::iterator* start )
+{
+  // set start pointer to the eprop history entry corresponding to time t1
+  if ( eprop_history_.empty() )
+  {
+    *start = eprop_history_.end();
+    return;
+  }
+  else
+  {
+    // compute the positions of the pointer pointing to the entry at t1
+    // since no time steps are missing in eprop history, just ensure that *start
+    // points at least to begin()
+    double t_first = eprop_history_.begin()->t_;
+    double h = Time::get_resolution().get_ms();
+    int pos_t1 = std::max( 0, ( ( int ) std::round( ( t1 - t_first ) / h ) ) );
+
+    std::deque< histentry_eprop >::iterator it_first = eprop_history_.begin();
+    *start = it_first + std::max( 0, pos_t1 );
+  }
+}
+
+
+void
+nest::EpropArchivingNode::register_update( double t_last_update, double t_current_update )
+{
+  // register spike time if it is not in the list, otherwise increase access counter
+  std::vector< histentry_extended >::iterator it_reg =
+    std::lower_bound( t_last_update_per_synapse_.begin(), t_last_update_per_synapse_.end(), t_current_update - eps_ );
+
+  if ( it_reg == t_last_update_per_synapse_.end() || fabs( t_current_update - it_reg->t_ ) > eps_ )
+  {
+    t_last_update_per_synapse_.insert( it_reg, histentry_extended( t_current_update, 0.0, 1 ) );
+  }
+  else
+  {
+    ++it_reg->access_counter_;
+  }
+  // search for old entry, decrease access counter, and delete entry if the access counter equals zero
+  it_reg =
+    std::lower_bound( t_last_update_per_synapse_.begin(), t_last_update_per_synapse_.end(), t_last_update - eps_ );
+
+  if ( it_reg == t_last_update_per_synapse_.end() || fabs( t_last_update - it_reg->t_ ) > eps_ )
+  {
+    std::cout << "found no entry to register in scanned " + std::to_string( t_last_update ) + " ms interval \n";
+  }
+  else
+  {
+    it_reg->access_counter_--;
+
+    if ( it_reg->access_counter_ == 0 )
+    {
+      it_reg = t_last_update_per_synapse_.erase( it_reg ); // erase old entry
+    }
+  }
+}
+
+
+void
+nest::EpropArchivingNode::get_spike_history( double t1,
+  double t2,
+  std::deque< double >::iterator* start,
+  std::deque< double >::iterator* finish )
+{
+  // set pointer to spike history entries corresponding to times t1 and t2
+  *finish = spike_history_.end();
+
+  if ( spike_history_.empty() )
+  {
+    *start = *finish;
+    return;
+  }
+  else
+  {
+    std::deque< double >::iterator runner1 =
+      std::lower_bound( spike_history_.begin(), spike_history_.end(), t1 + eps_ );
+
+    *start = runner1;
+
+    std::deque< double >::iterator runner2 = std::lower_bound( runner1, spike_history_.end(), t2 + eps_ );
+
+    *finish = runner2;
+  }
+}
+
+void
+nest::EpropArchivingNode::tidy_eprop_history()
+{
+  double earliest_time_to_keep = ( t_last_update_per_synapse_.begin() )->t_;
+  if ( !eprop_history_.empty() )
```

Comment:
```suggestion
  const double earliest_time_to_keep = ( t_last_update_per_synapse_.begin() )->t_;
  if ( not eprop_history_.empty() )
```

Reply:
This part of the code is refactored during which these two critical lines were removed.